### DISCUSSION TITLE
Behavior bindable property with DynamicResource not supported?

### DISCUSSION THREAD

**Discussion by Cybrosys** [dotnet/maui#13691] [Q&A]
Hi,

I raised this Q&A and also a bug in the MCT repo:
https://github.com/CommunityToolkit/Maui/discussions/1054

It is a question related to doing a `DynamicResource ` binding against a bindable property on the `StatusBarBehaior`, if it is supported or a bug. If it is a bug I think it belongs here in the main MAUI repo.

Here is a link to the bug in the MCT Repo:
https://github.com/CommunityToolkit/Maui/issues/1056

So, is this a bug in the MAUI repo or a limitation? For me as a developer and consumer, since the XAML does not show any error I would expect and assume that I would be able to use a `DynamicResource` binding for the `StatusBarBehavior`'s `StatusBarColor` bindable property.

I also expressed in one of my comments that if I wanted to start investigating this, where would I start? I do not know who would own the `DynamicResource` binding. When doing a normal `Bind` I would expect it to be owned by the UI element itself, targeting a defined `BindingContext`. However, `DynamicResource` bindings seem a bit more "magical" in that they work without a `BindingContext` being defined/assigned.

There are several issues currently related to `DynamicResource` bindings not working in .NET 7 but in .NET 6. So I do not know if this problem is related to `Behavor`s specifically or a core issue with `DynamicResource`.

**Update**

I got information here https://github.com/CommunityToolkit/Maui/issues/1056 that it can't use `DynamicResource` because it does not inherit from [VisualElement](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/VisualElement.cs#L770) or implement [IResourcesProvider](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/IResourcesProvider.cs).

I would argue that you should be able to use `DynamicResource` even when not inheriting from `VisualElement`. The reason being that you can put any type of value into a `ResourceDictionary`, so in essence it is just a normal binding where the source is the parent or app's `ResourceDictionary`.

**Comment by RsZoli** [comment]
well, after 2,5 years, i too would like to know about this!

**Comment by stephenquan** [comment]
On Mar 7, 2023 @TheCodeTraveler writes https://github.com/CommunityToolkit/Maui/issues/1056#issuecomment-1456498023:

> Behavior in .NET MAUI does not have support for ResourceDictionary and thus do not support StaticResource or DynamicResource
> 
> In .NET MAUI, [Behavior inherits from BindableObject](https://github.com/dotnet/maui/blob/599e4798d22306f7c06cd8f80d50b0d14bf6c841/src/Controls/src/Core/Interactivity/Behavior.cs#L10).
>
> For a class to support ResourceDictionary, it needs to either inherit from [VisualElement where public ResourceDictionary Resources { get; set; } exists](https://github.com/dotnet/maui/blob/599e4798d22306f7c06cd8f80d50b0d14bf6c841/src/Controls/src/Core/VisualElement.cs#L770), or implement [IResourceProvider](https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/IResourcesProvider.cs).

This explains why DynamicResource cannot be assigned to Behaviors, but it also suggests an interesting workaround. It works by assigning the DynamicResource to a property in a hidden VisualElement (e.g., Button.CommandParameter) and then extracting the StatusBarColor from it. For it to work, the VisualElement must be connected to see the merge dictionaries, hence the Parent line.

```xaml
<ContentPage x:Name="thisPage" ...>
    <ContentPage.Behaviors>
        <toolkit:StatusBarBehavior
           StatusBarColor="{Binding CommandParameter,
                                    x:DataType=Button,
                                    Source={Button CommandParameter={DynamicResource Tertiary},
                                                   Parent={Reference thisPage}}}" />
    </ContentPage.Behaviors>
</ContentPage>
```

Another solution is to write an alternate DynamicResource that carries its own VisualElement with a settable Parent, e.g.

```xaml
<ContentPage.Behaviors>
    <toolkit:StatusBarBehavior StatusBarColor="{local:DynamicResourceAlt Tertiary}" />
</ContentPage.Behaviors>
```

```c#
[ContentProperty(nameof(Key))]
[RequireService([typeof(IReferenceProvider), typeof(IProvideValueTarget)])]
public partial class DynamicResourceAlt : Button, IMarkupExtension<BindingBase>
{
    [BindableProperty] public partial string Key { get; set; } = string.Empty;
    public DynamicResourceAlt()
    {
        this.Parent = Application.Current;
        this.PropertyChanged += (s, e) =>
        {
            if (e.PropertyName == nameof(Key) && !string.IsNullOrEmpty(Key))
            {
                SetDynamicResource(Button.CommandParameterProperty, Key);
            }
        };
    }
    public object ProvideValue(IServiceProvider serviceProvider) => (this as IMarkupExtension<BindingBase>).ProvideValue(serviceProvider);
    BindingBase IMarkupExtension<BindingBase>.ProvideValue(IServiceProvider serviceProvider)
    {
        if (serviceProvider.GetService(typeof(IProvideValueTarget)) is IProvideValueTarget provideValueTarget && provideValueTarget.TargetObject is BindableObject targetObject)
        {
            this.SetBinding(BindingContextProperty, static (BindableObject t) => t.BindingContext, BindingMode.OneWay, source: targetObject);
        }
        return BindingBase.Create(static (DynamicResourceAlt t) => t.CommandParameter, BindingMode.OneWay, source: this);
    }
}
```

**Comment by stephenquan** [comment]
Suggestion, if we create a new `IBehavior` interface, we can change VisualElement.Behaviors as follows

```c#
public partial class VisualElement : NavigableElement, IAnimatable, IVisualElementController, IResourcesProvider, IStyleElement, IFlowDirectionController, IPropertyPropagationController, IVisualController, IWindowController, IView, IControlsVisualElement, IConstrainedView
{
    // public IList<Behavior> Behaviors
   public IList<IBehavior> Behaviors
   /* ... */
```

Then someone can subclass VisualElement and an IBehavior interface as a new Behavior, and this may "help" support DynamicResource. How we may utilize this could be 

```c#
//public partial class IconTintColorBehavior : Behavior<View>
public partial class IconTintColorBehavior : VisualElement, IBehavior<View>
{
    /* ... */
}
```

```xaml
<Image x:Name="globe" Source="globe_32x32.png">
    <Image.Behaviors>
        <toolkit:IconTintColorBehavior
           TintColor="{DynamicResource TargetColor}",
           Parent="{Reference globe}" />
    </Image.Behaviors>
</Image>
```

References:
 - https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/VisualElement/VisualElement.cs
 - Experimental PR (for discussion): https://github.com/dotnet/maui/pull/32350
